### PR TITLE
[#62] Integrate splash screen

### LINKED ISSUE #62: [Integrate] As a user, when I open the application I can see Splash screen
## Why
Implement integrations for Splash screen

## Acceptance Criteria
- As soon as application is launched, we can see `Splash screen`
- All UI has been integrated in UI ticket
- After sometime the application will be navigate to blank `Login screen`
**DO NOT IMPLEMENT LOGIN SCREEN**

## Resources
[Animations](https://www.figma.com/proto/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App)

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
Closes #62 

## What happened 👀

Implement integrations for Splash screen. 

- As soon as application is launched, we can see Splash screen
- All UI has been integrated in UI ticket
- After sometime the application will be navigate to blank Login screen
- DO NOT IMPLEMENT LOGIN SCREEN

## Insight 📝

- Implement ViewModel and logic to navigate from splash to login screen
- implement workflow to check instrumented tests

## Proof Of Work 📹

https://user-images.githubusercontent.com/32578035/184101676-659952ed-d255-4c41-97b8-16291c648071.mp4

### CODE REVIEW

**Comment by ryan-conway** [comment]
@kaungkhantsoe might need a rebase here 😉

File: app/google-services.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
+{
```
Comment (AVI5HEK) [inline app/google-services.json]: This file shouldn't be on VCS. Didn't we add it to `.gitignore`?

File: app/src/androidTest/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/splash/SplashScreenTest.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    val composeTestRule = createAndroidComposeRule<MainActivity>()
+
+    @Test
+    fun when_splash_screen_start_show_both_background_image_and_logo_image() {
+        composeTestRule.setContent {
+            NimbleSurveyJetpackComposeTheme {
+                SplashScreen(EmptyDestinationsNavigator)
+            }
+        }
+        composeTestRule.onNodeWithContentDescription(composeTestRule.activity.getString(R.string.splash_background_content_description))
```
Comment (AVI5HEK) [inline app/src/androidTest/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/splash/SplashScreenTest.kt]: Could we move `composeTestRule.activity.getString(...)` to a new method? 🤔 

```kotlin
private fun getString(resId: Int) : String = composeTestRule.activity.getString(resId)
```

File: app/src/androidTest/java/com/kks/nimblesurveyjetpackcompose/base/BaseAndroidComposeTest.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import androidx.compose.ui.test.junit4.createAndroidComposeRule
+import com.kks.nimblesurveyjetpackcompose.ui.presentation.main.MainActivity
+import org.junit.Rule
+
+abstract class BaseAndroidComposeTest {
+    @get:Rule
+    val composeTestRule = createAndroidComposeRule<MainActivity>()
+
+    fun getString(resId: Int) = composeTestRule.activity.getString(resId)
```
Comment (AVI5HEK) [inline app/src/androidTest/java/com/kks/nimblesurveyjetpackcompose/base/BaseAndroidComposeTest.kt]: Can make it `protected`.
```suggestion
    protected fun getString(resId: Int) = composeTestRule.activity.getString(resId)
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/splash/SplashScreen.kt [hunk 4]
```diff
@@ -32,15 +45,15 @@ fun SplashScreen() {
         Image(
             painter = painterResource(id = R.drawable.ic_logo_white),
             contentDescription = stringResource(id = R.string.splash_logo_content_description),
-            modifier = Modifier.size(201.0.dp, 48.0.dp)
+            modifier = Modifier
+                .size(201.0.dp, 48.0.dp)
```
Comment (AVI5HEK) [inline app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/splash/SplashScreen.kt]: The line break is back. 😆

**Review by AVI5HEK — APPROVED** [review]
The rest LGTM.

File: app/src/test/java/com/kks/nimblesurveyjetpackcompose/viewmodel/splash/SplashViewModelTest.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            viewModel.startTimerToNavigateToLogin()
+
+            var actual = false
+            launch {
+                delay(SPLASH_TIME)
+                actual = viewModel.shouldNavigateToLogin.value
+            }
+            advanceUntilIdle()
+
+            assertEquals(true, actual)
```
Comment (ryan-conway) [inline app/src/test/java/com/kks/nimblesurveyjetpackcompose/viewmodel/splash/SplashViewModelTest.kt:39]: Do we really need to declare a `var actual` for this? Why not directly access `shouldNavigateToLogin`? 🤔 
```suggestion
            assertEquals(true, viewModel.shouldNavigateToLogin.value)
```

File: app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/splash/SplashScreen.kt (hunk 4 above)
Comment (ryan-conway) [reply app/src/main/java/com/kks/nimblesurveyjetpackcompose/ui/presentation/splash/SplashScreen.kt]: @kaungkhantsoe Let's not forget this 😉

**Review by ryan-conway — COMMENTED** [review]
One more thing which we missed in the UI ticket tho 🤔 
On the [Figma Demo](https://www.figma.com/proto/GjRPOjDyZ6f4EDL3wKarRK/Challenge---Mobile-App?node-id=0%3A1&scaling=scale-down&page-id=0%3A1&starting-point-node-id=34%3A450), the Nimble logo fades in before the transition to login screen. One of the strengths of Jetpack Compose is how simple it is to create animations. How about we create another issue to add this in? :smile:

File: app/src/test/java/com/kks/nimblesurveyjetpackcompose/viewmodel/splash/SplashViewModelTest.kt (hunk 5 above)
Comment (kaungkhantsoe) [reply app/src/test/java/com/kks/nimblesurveyjetpackcompose/viewmodel/splash/SplashViewModelTest.kt:39]: @ryan-conway I am planning to make the transition animation after I have implemented the `Login UI`. BTW how did you implement that transition? I saw that Compose haven't officially support that animation yet. I saw [this experimental library](https://github.com/mxalbert1996/compose-shared-elements) though.

Comment (ryan-conway) [reply app/src/test/java/com/kks/nimblesurveyjetpackcompose/viewmodel/splash/SplashViewModelTest.kt:39]: You could combine `Splash` and `Login` into one screen and animate the composables inside it 😉

Comment (kaungkhantsoe) [reply app/src/test/java/com/kks/nimblesurveyjetpackcompose/viewmodel/splash/SplashViewModelTest.kt:39]: I see. I also had that thought but like you said I need to combine two screens and show and hide the other components.

**Comment by kaungkhantsoe** [comment]
@ryan-conway Updated 🚀

**Review by ryan-conway — APPROVED** [review]

**Pull request merged**